Add KEP-167 'metrics collector' (#246)
kubernetes-sigs/kube-scheduler-simulator: keps/167-metrics-collector/README.md

# KEP-167: Metrics Collector

## Summary

> Define a new Custom Resource `MetricsCollector` that will periodically collect metrics from the endpoint where metrics are published.

Reviewer: ```suggestion
Define a new Custom Resource `MetricsCollector` that periodically collects metrics from the endpoint where metrics are published.
```

`MetricsCollector` allows users to periodically collect arbitrary metrics exposed to them from a specified server.
This resource is primarily intended to be used in conjunction with the `Scenario` to benchmark simulators, however, it also provides a more general and simple functionality that can be used in other ways. (`Scenario` is the CR to run scenario-based simulation introduced in KEP-140.)

## Motivation
With `Scenario` introduced in KEP-140, we are now able to experiment with scheduling on a somewhat larger scale.
And you will surely want metrics against which you can measure the performance of your scheduler.
This KEP defines a new custom resource `MetricsCollector` to collect and record metrics.

### Goals 

> - Allow users to obtain scheduler metrics periodically.

Reviewer: ```suggestion
- Allow users to obtain metrics periodically.
```

> - Combined with `Scenario`, it allows users to record metrics during scenario execution.

Reviewer: It's kinda motivation, not goal itself. Remove it from here and move to somewhere else.

### Non-Goals

> - Allow users to compare benchmark results from one scheduler to another. (The scope of this proposal is to record(collect) only.)

Reviewer: It can be removed.

- See the result from Web UI. (maybe implemented in the future, but out of the scope of this proposal.)

## User Stories
### Story #1
A user has prepared a scheduler to be used for an internal product. 
They want to test the performance of the scheduler before putting it into production. 
This MetricsCollector will help the user to answer that question.

### Story #2
The user wants to tune the scheduler, and while Scenario allows the user to check the scheduler's performance in production to some extent, it does not tell the user how the scheduler is performing. 
By using the `MetricsCollector` together, users will have an indicator of appropriate tuning adjustments.

## Proposal

> The job MetricsCollector does is simple and straightforward.

Reviewer: ```suggestion
What MetricsCollector does is simple and straightforward.
```

> Once the MetricsCollector resource is created, it will continue to periodically retrieve metrics from the target scheduler endpoint.

Reviewer: ```suggestion
Once the MetricsCollector resource is created, it will continue to periodically retrieve metrics from the target endpoint.
```
